core/src/main/java/com/sequenceiq/cloudbreak/service/cluster/AmbariClusterService.java

```diff
@@ -487,4 +502,23 @@ private long getRemainingSpace(Map<String, Long> remainingNodes, Map<String, Map
         }
         return result;
     }
+
+    private Cluster doUpdateAmbariIp(Long clusterId, String ambariIp, int attempt) {
+        try {
+            Cluster cluster = clusterRepository.findById(clusterId);
+            cluster.setAmbariIp(ambariIp);
+            cluster = clusterRepository.save(cluster);
+            LOGGER.info("Updated cluster: [ambariIp: '{}'].", ambariIp);
+            return cluster;
+        } catch (OptimisticLockException | OptimisticLockingFailureException e) {
+            if (attempt <= MAX_RETRIES) {
+                LOGGER.info("Failed to update cluster's Ambari IP. [attempt: '{}', Cause: {}]. Trying to save it again.", attempt,
+                        e.getClass().getSimpleName());
+                return doUpdateAmbariIp(clusterId, ambariIp, attempt + 1);
+            } else {
+                throw new CloudbreakServiceException(String.format("Failed to update cluster '%s' in %d attempts. (while trying to update ambariIp)",
+                        clusterId, MAX_RETRIES), e);
+            }
+        }
+    }
```

Comment:
I don't think it's needed to save the cluster like this. It is always updated from one thread only.